File: cmd_mgr/windows/CmdMgr/ViewModels/ProcessManager.cs
using System.Diagnostics;
using System.Collections.Concurrent;
using System.Text;

namespace CmdMgr.ViewModels;

public enum ProcessStatus
{
    Running,
    Success,
    Failed,
    Stopped
}

public static class ProcessStatusExtensions
{
    public static string DisplayName(this ProcessStatus status) => status switch
    {
        ProcessStatus.Running => "Running",
        ProcessStatus.Success => "Succeeded",
        ProcessStatus.Failed => "Failed",
        ProcessStatus.Stopped => "Stopped",
        _ => status.ToString()
    };

    public static string Icon(this ProcessStatus status) => status switch
    {
        ProcessStatus.Running => "●",
        ProcessStatus.Success => "✅",
        ProcessStatus.Failed => "❌",
        ProcessStatus.Stopped => "⏹",
        _ => "○"
    };

    public static string BrushKey(this ProcessStatus status) => status switch
    {
        ProcessStatus.Running => "PrimaryBrush",
        ProcessStatus.Success => "SuccessBrush",
        ProcessStatus.Failed => "DangerBrush",
        ProcessStatus.Stopped => "WarningBrush",
        _ => "MutedBrush"
    };
}

/// <summary>
/// Information about a running (or recently completed) process.
/// </summary>
public class ProcessInfo : ViewModelBase
{
    private const int MaxLines = 5000;

    public int CommandId { get; }
    public Models.Command Command { get; }
    public Process Process { get; }
    public int HistoryId { get; }

    private string _output = "";
    public string Output
    {
        get => _output;
        set => SetField(ref _output, value);
    }

    private ProcessStatus _status = ProcessStatus.Running;
    public ProcessStatus Status
    {
        get => _status;
        set
        {
            if (SetField(ref _status, value))
                OnPropertyChanged(nameof(IsRunning));
        }
    }

    public bool IsRunning => Status == ProcessStatus.Running;

    public ProcessInfo(Models.Command command, Process process, int historyId)
    {
        Command = command;
        CommandId = command.Id;
        Process = process;
        HistoryId = historyId;
    }

    /// <summary>Appends text and trims to MaxLines to prevent unbounded memory growth.</summary>
    public void AppendOutput(string text)
    {
        var combined = _output + text;
        var lines = combined.Split('\n');
        Output = lines.Length > MaxLines
            ? string.Join("\n", lines[^MaxLines..])
            : combined;
    }
}

/// <summary>
/// Manages subprocess lifecycle on Windows.
/// </summary>
public class ProcessManager : ViewModelBase
{
    public ConcurrentDictionary<int, ProcessInfo> RunningProcesses { get; } = new();
    // Keeps last completed run per command so output remains viewable after exit.
    public ConcurrentDictionary<int, ProcessInfo> CompletedProcesses { get; } = new();

    private readonly ConcurrentDictionary<int, bool> _requestedStops = new();

    private string? _lastError;
    public string? LastError
    {
        get => _lastError;
        set => SetField(ref _lastError, value);
    }

    /// <summary>Raised whenever a command's running/completed process info changes.</summary>
    public event Action<int>? ProcessesChanged;

    public bool IsRunning(int commandId) =>
        RunningProcesses.TryGetValue(commandId, out var info) && info.IsRunning;

    public ProcessInfo? GetProcessInfo(int commandId) =>
        RunningProcesses.TryGetValue(commandId, out var r) ? r :
        CompletedProcesses.TryGetValue(commandId, out var c) ? c : null;

    public bool HasOutput(int commandId) =>
        !string.IsNullOrEmpty(GetProcessInfo(commandId)?.Output);

    public ProcessInfo? Start(Models.Command command, Models.Database database)
    {
        if (IsRunning(command.Id)) return null;

        int? historyId = null;
        try
        {
            Encoding encoding;
            try
            {
                encoding = Encoding.GetEncoding(
                    System.Globalization.CultureInfo.CurrentCulture.TextInfo.OEMCodePage);
            }
            catch
            {
                encoding = Encoding.UTF8;
            }

            var psi = new ProcessStartInfo
            {
                FileName = "cmd.exe",
                Arguments = $"/c {command.CommandText}",
                UseShellExecute = false,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                CreateNoWindow = true,
                StandardOutputEncoding = encoding,
                StandardErrorEncoding = encoding
            };

            if (!string.IsNullOrWhiteSpace(command.WorkingDirectory))
                psi.WorkingDirectory = command.WorkingDirectory;

            historyId = database.AddHistoryEntry(command.Id, DateTime.Now, "running");
            var proc = new Process { StartInfo = psi, EnableRaisingEvents = true };
            var info = new ProcessInfo(command, proc, historyId.Value);

            proc.OutputDataReceived += (_, e) =>
            {
                if (e.Data != null)
                    System.Windows.Application.Current?.Dispatcher.Invoke(
                        () => info.AppendOutput(e.Data + "\n"));
            };
            proc.ErrorDataReceived += (_, e) =>
            {
                if (e.Data != null)
                    System.Windows.Application.Current?.Dispatcher.Invoke(
                        () => info.AppendOutput($"[STDERR] {e.Data}\n"));
            };

            // Exited is the single source of truth for all cleanup and DB writes.
            proc.Exited += (_, _) =>
            {
                var wasRequestedStop = _requestedStops.TryRemove(command.Id, out _);
                var status = wasRequestedStop ? "terminated" : (proc.ExitCode == 0 ? "success" : "failed");
                System.Windows.Application.Current?.Dispatcher.Invoke(() =>
                {
                    var finalStatus = wasRequestedStop ? ProcessStatus.Stopped
                        : (proc.ExitCode == 0 ? ProcessStatus.Success : ProcessStatus.Failed);
                    info.Status = finalStatus;
                    command.Status = finalStatus;
                    database.UpdateHistoryEntry(info.HistoryId, DateTime.Now, status, info.Output);
                    RunningProcesses.TryRemove(command.Id, out _);
                    CompletedProcesses[command.Id] = info;
                    ProcessesChanged?.Invoke(command.Id);
                });
            };

            command.Status = ProcessStatus.Running;
            CompletedProcesses.TryRemove(command.Id, out _);
            RunningProcesses[command.Id] = info;
            ProcessesChanged?.Invoke(command.Id);

            if (!proc.Start())
            {
                command.Status = ProcessStatus.Failed;
                RunningProcesses.TryRemove(command.Id, out _);
                var message = $"Could not start \"{command.Name}\": process did not start.";
                info.Status = ProcessStatus.Failed;
                info.AppendOutput(message);
                CompletedProcesses[command.Id] = info;
                LastError = message;
                database.UpdateHistoryEntry(historyId.Value, DateTime.Now, "failed", message);
                ProcessesChanged?.Invoke(command.Id);
                return null;
            }

            proc.BeginOutputReadLine();
            proc.BeginErrorReadLine();
            return info;
        }
        catch (Exception ex)
        {
            command.Status = ProcessStatus.Failed;
            RunningProcesses.TryRemove(command.Id, out _);
            var message = $"Could not start \"{command.Name}\": {ex.Message}";
            LastError = message;
            if (historyId.HasValue)
                database.UpdateHistoryEntry(historyId.Value, DateTime.Now, "failed", message);
            ProcessesChanged?.Invoke(command.Id);
            return null;
        }
    }

    /// <summary>
    /// Sends a kill signal. All cleanup (IsRunning, DB update, dictionary removal)
    /// is handled exclusively by the Exited event handler to avoid race conditions.
    /// </summary>
    public void Stop(int commandId, Models.Database database)
    {
        if (!RunningProcesses.TryGetValue(commandId, out var info)) return;

        _requestedStops[commandId] = true;
        try
        {
            if (!info.Process.HasExited)
                info.Process.Kill(entireProcessTree: true);
        }
        catch { /* Process already exited – Exited handler will clean up */ }
    }

    public void StopAll(Models.Database database)
    {
        foreach (var cmdId in RunningProcesses.Keys)
            Stop(cmdId, database);
    }
}

File: cmd_mgr/macos/Sources/CmdMgr/ViewModels/ProcessManager.swift
import Foundation

/// Information about a running (or recently completed) process.
final class CommandProcess: ObservableObject, Identifiable {
    let commandId: Int
    let process: Process
    let historyId: Int

    @Published var output: String = ""
    @Published var isRunning: Bool = true

    private static let maxLines = 5000

    init(commandId: Int, process: Process, historyId: Int) {
        self.commandId = commandId
        self.process = process
        self.historyId = historyId
    }

    func appendOutput(_ text: String) {
        output.append(text)
        // Trim to avoid unbounded memory growth for long-running services.
        let lines = output.components(separatedBy: "\n")
        if lines.count > Self.maxLines {
            output = lines.suffix(Self.maxLines).joined(separator: "\n")
        }
    }
}

/// Manages subprocess lifecycle: start, stop, output capture.
final class ProcessManager: ObservableObject {

    @Published var runningProcesses: [Int: CommandProcess] = [:]
    // Keeps last completed run per command so output remains viewable after exit.
    @Published var completedProcesses: [Int: CommandProcess] = [:]

    private var requestedStops: Set<Int> = []

    func isRunning(_ commandId: Int) -> Bool {
        runningProcesses[commandId]?.isRunning ?? false
    }

    func hasOutput(_ commandId: Int) -> Bool {
        let info = runningProcesses[commandId] ?? completedProcesses[commandId]
        return !(info?.output.isEmpty ?? true)
    }

    func processInfo(_ commandId: Int) -> CommandProcess? {
        runningProcesses[commandId] ?? completedProcesses[commandId]
    }

    /// Start a command as a subprocess.
    func start(command: Command, database: Database) {
        guard !isRunning(command.id) else { return }

        let proc = Process()
        proc.executableURL = URL(fileURLWithPath: "/bin/sh")
        proc.arguments = ["-c", command.command]
        if let wd = command.workingDirectory, !wd.isEmpty {
            proc.currentDirectoryURL = URL(fileURLWithPath: wd)
        }

        let outPipe = Pipe()
        let errPipe = Pipe()
        proc.standardOutput = outPipe
        proc.standardError = errPipe

        let historyId = database.addHistoryEntry(
            commandId: command.id, startTime: Date(), status: "running")

        let info = CommandProcess(commandId: command.id, process: proc, historyId: historyId)

        outPipe.fileHandleForReading.readabilityHandler = { [weak info] handle in
            let data = handle.availableData
            guard !data.isEmpty, let str = String(data: data, encoding: .utf8) else { return }
            DispatchQueue.main.async { info?.appendOutput(str) }
        }

        errPipe.fileHandleForReading.readabilityHandler = { [weak info] handle in
            let data = handle.availableData
            guard !data.isEmpty, let str = String(data: data, encoding: .utf8) else { return }
            DispatchQueue.main.async { info?.appendOutput("[STDERR] \(str)") }
        }

        proc.terminationHandler = { [weak self, weak info] proc in
            guard let self = self, let info = info else { return }
            outPipe.fileHandleForReading.readabilityHandler = nil
            errPipe.fileHandleForReading.readabilityHandler = nil
            let remainOut = outPipe.fileHandleForReading.readDataToEndOfFile()
            let remainErr = errPipe.fileHandleForReading.readDataToEndOfFile()

            DispatchQueue.main.async {
                let wasRequestedStop = self.requestedStops.remove(info.commandId) != nil
                let status = wasRequestedStop ? "terminated"
                    : (proc.terminationStatus == 0 ? "success" : "failed")
                if let s = String(data: remainOut, encoding: .utf8), !s.isEmpty {
                    info.appendOutput(s)
                }
                if let s = String(data: remainErr, encoding: .utf8), !s.isEmpty {
                    info.appendOutput("[STDERR] \(s)")
                }
                info.isRunning = false
                database.updateHistoryEntry(
                    id: info.historyId, endTime: Date(), status: status, output: info.output)
                self.runningProcesses.removeValue(forKey: info.commandId)
                self.completedProcesses[info.commandId] = info
            }
        }

        do {
            completedProcesses.removeValue(forKey: command.id)
            runningProcesses[command.id] = info
            try proc.run()
        } catch {
            runningProcesses.removeValue(forKey: command.id)
            database.updateHistoryEntry(
                id: historyId, endTime: Date(), status: "failed",
                output: "Launch error: \(error.localizedDescription)")
        }
    }

    /// Stop a running command. Only sends signals; cleanup is done in terminationHandler.
    func stop(commandId: Int, database: Database) {
        guard let info = runningProcesses[commandId] else { return }
        requestedStops.insert(commandId)
        info.process.terminate()  // SIGTERM – give process a chance to clean up

        // Escalate to SIGKILL after 3 s if still alive.
        DispatchQueue.global().asyncAfter(deadline: .now() + 3) { [weak info] in
            guard let info = info, info.process.isRunning else { return }
            Foundation.kill(Int32(info.process.processIdentifier), SIGKILL)
        }
    }

    /// Stop all running processes.
    func stopAll(database: Database) {
        for cmdId in runningProcesses.keys {
            stop(commandId: cmdId, database: database)
        }
    }

    /// Blocks the calling (background) thread until all processes in `infos` have exited,
    /// then waits briefly for terminationHandler dispatches to drain on the main queue.
    func waitForAll(processes: [CommandProcess]) {
        for info in processes {
            // Safety kill after 3 s in case SIGTERM was ignored.
            DispatchQueue.global().asyncAfter(deadline: .now() + 3) { [weak info] in
                guard let info = info, info.process.isRunning else { return }
                Foundation.kill(Int32(info.process.processIdentifier), SIGKILL)
            }
            info.process.waitUntilExit()
        }
        // Allow terminationHandler's DispatchQueue.main.async closures to flush.
        Thread.sleep(forTimeInterval: 0.2)
    }
}
